plugin/output/s3/s3.go

```diff
@@ -96,97 +217,167 @@ func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginP
 	p.controller = params.Controller
 	p.logger = params.Logger
 	p.config = config.(*Config)
+	p.params = params
 
-	// set up compression
+	// outPlugCount is defaultBucket + multi_buckets count, use to set maps size.
+	outPlugCount := len(p.config.MultiBuckets) + 1
+
+	// set up compression.
 	newCompressor, ok := compressors[p.config.CompressionType]
 	if !ok {
 		p.logger.Fatalf("compression type: %s is not supported", p.config.CompressionType)
 	}
 	p.compressor = newCompressor(p.logger)
 
-	dir, f := filepath.Split(p.config.FileConfig.TargetFile)
-
-	p.targetDir = dir
-	p.fileExtension = filepath.Ext(f)
-	p.fileName = f[0 : len(f)-len(p.fileExtension)]
+	// dir for all bucket files.
+	targetDirs := p.getDirs(outPlugCount)
+	// file for each bucket.
+	fileNames := p.getFileNames(outPlugCount)
 
-	p.uploadCh = make(chan string, p.config.FileConfig.WorkersCount_*4)
-	p.compressCh = make(chan string, p.config.FileConfig.WorkersCount_)
+	p.uploadCh = make(chan fileDTO, p.config.FileConfig.WorkersCount_*4)
+	p.compressCh = make(chan fileDTO, p.config.FileConfig.WorkersCount_)
 
 	for i := 0; i < p.config.FileConfig.WorkersCount_; i++ {
 		longpanic.Go(p.uploadWork)
 		longpanic.Go(p.compressWork)
 	}
 
-	// initialize minio client object.
-	minioClient, err := minio.New(p.config.Endpoint, p.config.AccessKey, p.config.SecretKey, p.config.Secure)
-	if err != nil || minioClient == nil {
+	// initialize minio clients.
+	defaultClient, clients, err := p.minioClients()
+	if err != nil {
 		p.logger.Panicf("could not create minio client, error: %s", err.Error())
 	}
-	p.client = minioClient
+	p.defaultClient = defaultClient
+	p.clients = clients
+
+	err = p.startPlugins(params, outPlugCount, targetDirs, fileNames)
+	switch err.(type) {
+	case nil:
+		break
+	case *ErrCreateOutputPluginPanic:
+		p.logger.Panic(err.Error())
+	case *ErrCreateOutputPluginNoSuchBucket:
+		p.logger.Fatal(err.Error())
+	}
+
+	p.uploadExistingFiles(targetDirs, fileNames)
+}
+
+func (p *Plugin) Stop() {
+	p.outPlugins.Stop()
+}
+
+func (p *Plugin) Out(event *pipeline.Event) {
+	p.outPlugins.Out(event, pipeline.PluginSelector{
+		CondType:  pipeline.ByNameSelector,
+		CondValue: p.decideReceiverBucket(event),
+	})
+}
+
+// decideReceiverBucket decide which s3 bucket shall receive event.
+func (p *Plugin) decideReceiverBucket(event *pipeline.Event) string {
+	bucketName := event.Root.Dig(p.config.BucketFieldInEvent).AsString()
+	// no BucketFieldInEvent in message, it's DefaultBucket, showtime.
+	if len(bucketName) == 0 {
+		return p.config.DefaultBucket
+	}
+	// Bucket exists.
+	if p.outPlugins.Exists(bucketName) {
+		return bucketName
+	}
+
+	// To avoid concurrent creation of bucketName plugin add take lock.
+	p.dynamicPlugCreationMu.Lock()
+	defer p.dynamicPlugCreationMu.Unlock()
```

Comment:
may be move this locks and the check afterwards inside `p.tryRunNewPlugin(bucketName)`, so the lock will be inside and nobody will forget about taking it

Reply:
Good point.